### ISSUE TITLE
UUID type primary key

### ISSUE THREAD

**Issue by asif-mahmud** [pyeve/eve-sqlalchemy#156]
Hi, my setup is PostgreSQL+Psycopg2, Eve 0.7, Flask 0.12. I wanted to use UUID as primary key. So I followed this http://python-eve.org/tutorials/custom_idfields.html#handling-uuid-fields and added the following column in my models -

```
id = Column(postgresql.UUID(as_uuid=True), primary_key=True, server_default=func.uuid_generate_v4())
```
My database is being upgraded (by alembic) and i checked it.
Followed the tutorial i linked up before. One additional thing that i did is instead of subclassing mongo validator i subclassed ValidatorSQL.
But now when i run the application and try to POST a new entry without any id parameter it fails and says `{"id": "required field"}`. 
Why is it complaining about id, when i have configured the server to insert id automatically? I also tried using `on_insert` event hook of `Eve`, but the event isn't firing up at all!

Please help.
Thanks in advance.

**Comment by dkellner** [comment]
One first guess is: You're using `server_default` and not `default`, which is used to generate the schema configuration in your Eve `DOMAIN` dict. So I'd bet your schema for that field is missing a `default` key and therefore the validator says you need one. And as it's a primary key, it's of course required.

Can you try to manually set `DOMAIN['endpoint']['schema']['id']['required'] = False`?

**Comment by asif-mahmud** [comment]
Hi, I have did what you suggested (writing `'required' = False` explicitly). That worked. 

Thanks a lot.

I have been making a project template combining **Eve, Eve-SQLAlchemy, Flask-Migrate** together. It's here- https://github.com/asif-mahmud/eve_runner . It has a few features which may be useful for developer with such setup.

I was wondering, if it is possible to mark the `Column`s with `server_default` option as `required=False` when rendering the schema. It would be really helpful. I would like to work to implement this feature here. Let me know what you think.

**Comment by dkellner** [comment]
It would be great if you can implement it! One thing you have to keep in mind is, that `server_default` cannot be handled before the database level. But this is what Eve usually does with default fields. So when we add this we should make this clear in the documentation, too. And write some integration tests for all HTTP methods.

So  I would start adding a new integration test case in `tests/integration/` and add another test to `tests/config/resourceconfig/schema.py`. The implementation should actually be quite simple, `fieldconfig.py` should be the place to start.

**Comment by asif-mahmud** [comment]
Hi, looking at the SQLAlchemy docs, i came up with the following code to detect any `server_default` clause in any column of any table -

```
from sqlalchemy import inspect
    for t in db.Model.metadata.tables:
        table = db.Model.metadata.tables[t]
        # insp = inspect(table)        
        # columns = insp.get_columns(table)
        for c in table.columns:
            sd = getattr(c, 'server_default')
            if sd:  # column has a server_default clause
                app.logger.info('{}:{}'.format(c, sd))
```
here the `app` is `Eve` application. Of course i imported all the modules that have model defined in them first.

**Comment by asif-mahmud** [comment]
Hi, looking at the SQLAlchemy docs, i came up with the following code to detect any `server_default` clause in any column of any table -

```
from sqlalchemy import inspect
    for t in db.Model.metadata.tables:
        table = db.Model.metadata.tables[t]
        # insp = inspect(table)        
        # columns = insp.get_columns(table)
        for c in table.columns:
            sd = getattr(c, 'server_default')
            if sd:  # column has a server_default clause
                app.logger.info('{}:{}'.format(c, sd))
```
here the `app` is `Eve` application. Of course i imported all the modules that have model defined in them first.

**Comment by asif-mahmud** [comment]
So there are 2 things to keep in mind -
1. All SQLAlchemy models must be imported somehow first
2. Before generating Eve schema, all models must be looked out to see if there are columns with server_default clause. Those columns should be made `not required` to Eve application.

How to do suggest to do the first one? 

PS: the project template that i have talked about before has some functionality - loading all models first then registering schema for Eve application. Did it in a different way than the Eve documented way. It requires 2 app configuration options- `MODEL_DIRS` and `MODEL_EXCLUDE_FILES`.

**Comment by dkellner** [comment]
The way configuration works it is ensured that the models you will refer to in your `settings.py` have to be imported there. If this is done by hand or using an automatism as in your project template should not matter for this case.

And there shouldn't be a need for  "manually" walking through all models and look for `server_default` - just take a look at how `default` is currently implemented in `fieldconfig.py`. So referring to your code snippet above: you can save yourself from looping over models and columns, you will get a `FieldConfig` object for each column and can handle `server_default` there.

**Comment by dkellner** [comment]
Handling `server_default` was addressed in PR #160.
